oidc.go

```diff
@@ -0,0 +1,196 @@
+package headscale
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/gin-gonic/gin"
+	"github.com/patrickmn/go-cache"
+	"github.com/rs/zerolog/log"
+	"golang.org/x/oauth2"
+	"gorm.io/gorm"
+	"net/http"
+	"time"
+)
+
+type IDTokenClaims struct {
+	Name     string   `json:"name,omitempty"`
+	Groups   []string `json:"groups,omitempty"`
+	Email    string   `json:"email"`
+	Username string   `json:"preferred_username,omitempty"`
+}
+
+var oidcProvider *oidc.Provider
+var oauth2Config *oauth2.Config
+var stateCache *cache.Cache
```

Comment:
Could you move these three from global into the `Headscale` object?

Reply:
And maybe the stateCache can be named oidcMachineKeyCache or something explicit.